src/Course/FileIO.hs

```diff
@@ -81,50 +82,50 @@ the contents of c
 
 -- Given the file name, and file contents, print them.
 -- Use @putStrLn@.
-printFile ::
-  FilePath
-  -> Chars
-  -> IO ()
-printFile =
-  error "todo: Course.FileIO#printFile"
+printFile :: FilePath -> Chars -> IO ()
+printFile path chars = putStrLn ("The path: " ++  path ++ ". The chars: " ++ chars)
 
 -- Given a list of (file name and file contents), print each.
 -- Use @printFile@.
-printFiles ::
-  List (FilePath, Chars)
-  -> IO ()
-printFiles =
-  error "todo: Course.FileIO#printFiles"
+printFiles :: List (FilePath, Chars) -> IO ()
+printFiles = voidSequenceMap $ uncurry printFile
 
 -- Given a file name, return (file name and file contents).
 -- Use @readFile@.
-getFile ::
-  FilePath
-  -> IO (FilePath, Chars)
-getFile =
-  error "todo: Course.FileIO#getFile"
+getFile :: FilePath -> IO (FilePath, Chars)
+getFile path = readFile path >>= \content -> return (path, content)
+-- getFile path = do
+--   content <- readFile path
+--   return (path, content)
+
 
 -- Given a list of file names, return list of (file name and file contents).
 -- Use @getFile@.
-getFiles ::
-  List FilePath
-  -> IO (List (FilePath, Chars))
-getFiles =
-  error "todo: Course.FileIO#getFiles"
+getFiles :: List FilePath -> IO (List (FilePath, Chars))
+getFiles = sequenceMap getFile
```

Comment:
that's really nice :)

Reply:
This is the stuff that I find it really difficult to express to Elm peeps, generally... that there's this way of writing code where you build up tiny functions and compose them, and the compositions are "tiny" and you're kind of layering a stack... and you get the best parts of simplicity (in layering) and that gives you the ability to deal with complex things in a simple way.

I feel like *that* is the kind of simplicity that will let us attack and deal with the huge complexity in programming that we usually end up having to deal with. And you can do this in any language when you understand it... Haskell just happens to make it difficult *not* to do it :) or at least... slightly more difficult.

Have you seen this ruby library? https://github.com/soveran/syro gosh... it's nice. This guy did a bunch of elm and maybe even haskell, learnt it and understood it well, and then applied that kind of compositional thinking to his Ruby library building.

Reply:
Yeah I think you're right. Elm has different goals — it aims to be easy to use, I believe. In a way that's actually antithetical to abstraction ;-)

Reply:
I always end up wondering if abstraction can be layered down to concretion... so that it's easy for beginners, but when you look into the concrete thing, it's actually just a super abstract that that's been specialised... or at least having separated layers whereby if you open one of the layers up, you get more abstract stuff, but you need to understand more stuff to get it.

I also wonder if the language couldn't actually teach you this, and unlock the layers when you're ready. Just ponderings I often have :)

Reply:
There's a synthesizer program I use which is called Reaktor... it has these layers... on the outside, you can get a synth built with Reaktor and just use it like an application, but if you crack open the "virtual case" so to speak, you can see the components its made up of... these are things that are visually wired together into a circuit... then you're operating at the modular synthesizer designer level... and then if you want to look into what makes up each component, you find the you can open the *third* layer... to find (I think they're called) elements: the base building blocks... and that's at a pretty low level (like... physical sound engineer level). 

I like that architecture.

Reply:
in regards to the generalisation, that's how you do pure mathematics right? you start with definitions and axioms, which might be considered abstractions, and you start building to more concrete stuff from there. Maybe if pure mathematics were more trendy that would be easier to apply.